### PR TITLE
Fix msan issue in keyed siphash

### PR DESCRIPTION
**Pull request by rschu1ze**
[Issue](https://s3.amazonaws.com/clickhouse-test-reports/0/ffdd91669471f4934704f98f0191524496b4e85b/fuzzer_astfuzzermsan/report.html)

Repro:
`SELECT hex(sipHash128ReferenceKeyed((toUInt64(2), toUInt64(-9223372036854775807)))) GROUP BY (toUInt64(506097522914230528), toUInt64(now64(2, NULL + NULL), 1084818905618843912)), toUInt64(2), NULL + NULL, char(-2147483649, 1)`

Minimal repro:
`SELECT sipHash64Keyed((2::UInt64, toUInt64(2)), 4) GROUP BY toUInt64(2)`

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix invalid memory access with keyed siphash functions.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit ad4a21034f79ba296153ea153a34d8829e7d1f6b with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50245/ad4a21034f79ba296153ea153a34d8829e7d1f6b/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Functions/FunctionsHashing.h [hunk 1]
```diff
@@ -91,12 +91,18 @@ namespace impl
             throw Exception(ErrorCodes::NOT_IMPLEMENTED, "wrong tuple size: key must be a tuple of 2 UInt64");
 
         if (const auto * key0col = checkAndGetColumn<ColumnUInt64>(&(tuple->getColumn(0))))
-            ret.key0 = key0col->get64(0);
+        {
```
Comment (alexey-milovidov) [inline src/Functions/FunctionsHashing.h:94]: Interesting, I don't understand the modification.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@rschu1ze I didn't understand the change.

**Comment by rschu1ze** [comment]
TLDR; No need to roll back #45513. This is most likely a false alarm.

The issue reproduced reliably on my machine with build `cmake -DCMAKE_BUILD_TYPE=None -DSANITIZE=memory -S . -B build` and queries
- `SELECT hex(sipHash128ReferenceKeyed((toUInt64(2), toUInt64(-9223372036854775807)))) GROUP BY (toUInt64(506097522914230528), toUInt64(now64(2, NULL + NULL), 1084818905618843912)), toUInt64(2), NULL + NULL, char(-2147483649, 1);`
- `SELECT sipHash64Keyed((2::UInt64, toUInt64(2)), 4) GROUP BY toUInt64(2);` (minimized former repro)

(look for `==3318566==WARNING: MemorySanitizer: use-of-uninitialized-value` in the server log)

I then looked how reading from `ColumnUInt64` works in other places and to my surprise, that fixed the problem.

The difference is:
- `ColumnVector<UInt64>::get64()` (used before) calls `bit_cast` which does a `memcpy`.
- `PodArray::operator[]` (now) returns the value directly.

There is nothing wrong with either path. I think msan simply got confused here. Actually, `get64` is used very rarely in the codebase, otherwise other issues would have shown up elsewhere.

Now, a new msan issue poped up in CI for query `SELECT hex(sipHash64Keyed((toUInt64(9223372036854775806), toUInt64(-9223372036854775808)), char(2147483646, -2147483648, 1, 3, 4, 7, 2147483647))) GROUP BY toUInt64(257), (toUInt64(9223372036854775806), toUInt64(2147483646))`. I'll look into it.

**Comment by rschu1ze** [comment]
> I'll try again as soon as I get the courage to rebuild everything.

Nevermind. I now found the true problem which was that `parseSipHash` didn't care if the chunk is empty or not. In the latter case, it read uninitialized memory.

File: src/Functions/FunctionsHashing.h [hunk 2]
```diff
@@ -1450,7 +1441,7 @@ class FunctionAnyHash : public IFunction
 
         KeyType key{};
         if constexpr (Keyed)
-            if (!arguments.empty())
+            if (!arguments.empty() && input_rows_count != 0)
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/FunctionsHashing.h]: Yepp, that should work. Done as proposed, thanks!

**Comment by tavplubix** [comment]
Looks like it breaks AST Fuzzers and `02534_keyed_siphash` with new analyzer
